### PR TITLE
Fix tiflash crash in `ThreadedWorker`

### LINKED ISSUE #7860: TiFlash Node Crash in `ThreadedWorker`
```
[2023/07/28 14:17:19.589 +08:00] [INFO] [common_checker.go:187] ["debug tiflash end search log"] ["end time"=2023/07/28 14:17:05.312 +08:00] [res="[\"[2023/07/28 14:16:39.834 +08:00] [Error] [BaseDaemon.cpp:368] [\\\"(from thread 582) Terminate called after throwing an instance of DB::Exception\\\
Code: 0, e.displayText() = DB::Exception: Assert isCancelled() fail!, e.what() = DB::Exception\\\
Stack trace:\\\
\\\
\\\
       0x75a212d\\\\tterminate_handler() [tiflash+123347245]\\\
                \\\\tlibs/libdaemon/src/BaseDaemon.cpp:633\\\
  0x7ff244420a13\\\\tstd::__terminate(void (*)()) [libc++abi.so.1+236051]\\\
  0x7ff2444209b8\\\\tstd::terminate() [libc++abi.so.1+235960]\\\
       0x1d27507\\\\tvoid DB::exception_details::logAndTerminate<>(char const*, int, char const*, std::__1::shared_ptr<DB::Logger> const&) [tiflash+30569735]\\\
                \\\\tdbms/src/Common/Exception.h:220\\\
       0x76d688b\\\\tstd::__1::__function::__func<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask> >::startInBackground()::'lambda'()::operator()() const::'lambda0'(), std::__1::allocator<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentT\\\"] [source=BaseDaemon] [thread_id=1208]\
\",\"[2023/07/28 14:16:40.400 +08:00] [Error] [BaseDaemon.cpp:368] [\\\"(from thread 511) Terminate called after throwing an instance of DB::Exception\\\
Code: 0, e.displayText() = DB::Exception: Assert isCancelled() fail!, e.what() = DB::Exception\\\
Stack trace:\\\
\\\
\\\
       0x75a212d\\\\tterminate_handler() [tiflash+123347245]\\\
                \\\\tlibs/libdaemon/src/BaseDaemon.cpp:633\\\
  0x7f93bc52ca13\\\\tstd::__terminate(void (*)()) [libc++abi.so.1+236051]\\\
  0x7f93bc52c9b8\\\\tstd::terminate() [libc++abi.so.1+235960]\\\
       0x1d27507\\\\tvoid DB::exception_details::logAndTerminate<>(char const*, int, char const*, std::__1::shared_ptr<DB::Logger> const&) [tiflash+30569735]\\\
                \\\\tdbms/src/Common/Exception.h:220\\\
       0x76d688b\\\\tstd::__1::__function::__func<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask> >::startInBackground()::'lambda'()::operator()() const::'lambda0'(), std::__1::allocator<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentT\\\"] [source=BaseDaemon] [thread_id=1056]\
\",\"[2023/07/28 14:16:40.391 +08:00] [Error] [BaseDaemon.cpp:368] [\\\"(from thread 228) Terminate called after throwing an instance of DB::Exception\\\
Code: 0, e.displayText() = DB::Exception: Assert isCancelled() fail!, e.what() = DB::Exception\\\
Stack trace:\\\
\\\
\\\
       0x75a212d\\\\tterminate_handler() [tiflash+123347245]\\\
                \\\\tlibs/libdaemon/src/BaseDaemon.cpp:633\\\
  0x7efda304ea13\\\\tstd::__terminate(void (*)()) [libc++abi.so.1+236051]\\\
  0x7efda304e9b8\\\\tstd::terminate() [libc++abi.so.1+235960]\\\
       0x1d27507\\\\tvoid DB::exception_details::logAndTerminate<>(char const*, int, char const*, std::__1::shared_ptr<DB::Logger> const&) [tiflash+30569735]\\\
                \\\\tdbms/src/Common/Exception.h:220\\\
       0x76d688b\\\\tstd::__1::__function::__func<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask> >::startInBackground()::'lambda'()::operator()() const::'lambda0'(), std::__1::allocator<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentT\\\"] [source=BaseDaemon] [thread_id=1063]\
\"]"]
[2023/07/28 14:17:19.589 +08:00] [ERROR] [common_checker.go:215] ["tiflash crash"] [log="[\"[2023/07/28 14:16:39.834 +08:00] [Error] [BaseDaemon.cpp:368] [\\\"(from thread 582) Terminate called after throwing an instance of DB::Exception\\\
Code: 0, e.displayText() = DB::Exception: Assert isCancelled() fail!, e.what() = DB::Exception\\\
Stack trace:\\\
\\\
\\\
       0x75a212d\\\\tterminate_handler() [tiflash+123347245]\\\
                \\\\tlibs/libdaemon/src/BaseDaemon.cpp:633\\\
  0x7ff244420a13\\\\tstd::__terminate(void (*)()) [libc++abi.so.1+236051]\\\
  0x7ff2444209b8\\\\tstd::terminate() [libc++abi.so.1+235960]\\\
       0x1d27507\\\\tvoid DB::exception_details::logAndTerminate<>(char const*, int, char const*, std::__1::shared_ptr<DB::Logger> const&) [tiflash+30569735]\\\
                \\\\tdbms/src/Common/Exception.h:220\\\
       0x76d688b\\\\tstd::__1::__function::__func<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask> >::startInBackground()::'lambda'()::operator()() const::'lambda0'(), std::__1::allocator<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentT\\\"] [source=BaseDaemon] [thread_id=1208]\
\",\"[2023/07/28 14:16:40.400 +08:00] [Error] [BaseDaemon.cpp:368] [\\\"(from thread 511) Terminate called after throwing an instance of DB::Exception\\\
Code: 0, e.displayText() = DB::Exception: Assert isCancelled() fail!, e.what() = DB::Exception\\\
Stack trace:\\\
\\\
\\\
       0x75a212d\\\\tterminate_handler() [tiflash+123347245]\\\
                \\\\tlibs/libdaemon/src/BaseDaemon.cpp:633\\\
  0x7f93bc52ca13\\\\tstd::__terminate(void (*)()) [libc++abi.so.1+236051]\\\
  0x7f93bc52c9b8\\\\tstd::terminate() [libc++abi.so.1+235960]\\\
       0x1d27507\\\\tvoid DB::exception_details::logAndTerminate<>(char const*, int, char const*, std::__1::shared_ptr<DB::Logger> const&) [tiflash+30569735]\\\
                \\\\tdbms/src/Common/Exception.h:220\\\
       0x76d688b\\\\tstd::__1::__function::__func<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask> >::startInBackground()::'lambda'()::operator()() const::'lambda0'(), std::__1::allocator<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentT\\\"] [source=BaseDaemon] [thread_id=1056]\
\",\"[2023/07/28 14:16:40.391 +08:00] [Error] [BaseDaemon.cpp:368] [\\\"(from thread 228) Terminate called after throwing an instance of DB::Exception\\\
Code: 0, e.displayText() = DB::Exception: Assert isCancelled() fail!, e.what() = DB::Exception\\\
Stack trace:\\\
\\\
\\\
       0x75a212d\\\\tterminate_handler() [tiflash+123347245]\\\
                \\\\tlibs/libdaemon/src/BaseDaemon.cpp:633\\\
  0x7efda304ea13\\\\tstd::__terminate(void (*)()) [libc++abi.so.1+236051]\\\
  0x7efda304e9b8\\\\tstd::terminate() [libc++abi.so.1+235960]\\\
       0x1d27507\\\\tvoid DB::exception_details::logAndTerminate<>(char const*, int, char const*, std::__1::shared_ptr<DB::Logger> const&) [tiflash+30569735]\\\
                \\\\tdbms/src/Common/Exception.h:220\\\
       0x76d688b\\\\tstd::__1::__function::__func<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask> >::startInBackground()::'lambda'()::operator()() const::'lambda0'(), std::__1::allocator<DB::ThreadedWorker<std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentTask>, std::__1::shared_ptr<DB::DM::Remote::RNReadSegmentT\\\"] [source=BaseDaemon] [thread_id=1063]\
\"]"] ["end time"=2023/07/28 14:17:05.312 +08:00] 
```

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #7860

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by guo-shaoge — APPROVED** [review]

File: dbms/src/Common/ThreadedWorker.h [hunk 1]
```diff
@@ -185,7 +185,7 @@ class ThreadedWorker
                         // Case B: There is something wrong with the downstream
                         // In case B, we need to populate the error to upstream, so that the whole
                         // pipeline is cancelled.
-                        auto cancel_reason = source_queue->getCancelReason();
+                        auto cancel_reason = result_queue->getCancelReason();
```
Comment (JaySon-Huang) [inline dbms/src/Common/ThreadedWorker.h:188]: good catch! I'll push a commit to add a ut about it

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**